datastream-server/src/main/java/com/linkedin/datastream/server/zk/ZkAdapter.java

```diff
@@ -921,6 +938,70 @@ public void waitForDependencies(DatastreamTaskImpl task, Duration timeout) {
     });
   }
 
+  /**
+   * Clean up partition movement info for a particular datastream
+   * @param connectorType Connector name
+   * @param datastreamGroupName the name of datastream group
+   */
+  public void cleanUpPartitionMovement(String connectorType, String datastreamGroupName) {
+    String path = KeyBuilder.getTargetAssignment(_cluster, connectorType, datastreamGroupName);
+    if (_zkclient.exists(path)) {
+      _zkclient.deleteRecursive(path);
+    }
+    LOG.info("clean up Target assignment info for {}", datastreamGroupName);
+  }
+
+  /**
+   * Get the datastream group which contains the partition movement info
+   * @param connectorType Connector name
+   */
+  public List<String> getDatastreamsNeedPartitionMovement(String connectorType) {
+    String path = KeyBuilder.getTargetAssignmentBase(_cluster, connectorType);
+    if (_zkclient.exists(path)) {
+      return _zkclient.getChildren(path);
+    }
+    return Collections.emptyList();
+  }
+
+  /**
+   * Get a partition movement info for a particular datastream
+   * @param connectorType Connector name
+   * @param datastreamGroupName the name of datastream group
+   * @return The target assignment mapping information with the instance name -> topicPartition
+   */
+  public Map<String, Set<String>> getPartitionMovement(String connectorType, String datastreamGroupName) {
+    Map<String, Set<String>> result = new HashMap<>();
+
+    String path = KeyBuilder.getTargetAssignment(_cluster, connectorType, datastreamGroupName);
+    if (_zkclient.exists(path)) {
+      List<String> nodes = _zkclient.getChildren(path);
+      Collections.sort(nodes);
+      for (String node : nodes) {
+        String content = _zkclient.readData(path + '/' + node);
+        TargetAssignment assignment = TargetAssignment.fromJson(content);
+
+        //Compute the correct instance
+        Map<String, String> hostInstanceMap = new HashMap<>();
+        if (_liveInstancesProvider.getLiveInstances() != null) {
+          _liveInstancesProvider.getLiveInstances().forEach(instance -> {
+            try {
+              hostInstanceMap.put(instance.substring(0, instance.lastIndexOf('-')), instance);
+            } catch (Exception ex) {
+              LOG.info("Fail to parse instance name {}", instance, ex);
+            }
+          });
+        }
+        if (hostInstanceMap.containsKey(assignment.getTargetHost()) && assignment.getPartitionNames() != null) {
+          LOG.info("Added assignment {} {}", assignment.getTargetHost(), assignment.getPartitionNames());
+          result.put(hostInstanceMap.get(assignment.getTargetHost()), new HashSet<>(assignment.getPartitionNames()));
```

Comment:
What happens when there are two requests to move two different partitions sets to the same host?

Reply:
I sorted them based on the timestamps, so the last write win. I've added a doc on that behaviorl

Reply:
Should we process all of them instead?

Reply:
Yes, all of them will be processed, but the last write wins.

Reply:
I changed the logic so that we will process all of them. ex. host1 -> p1, p2 and then host -> p2, p3 it will become host 1 -> p1, p2, p3. Have also added a test.